### PR TITLE
[WIP] [Feature] Financial cashflow forecast

### PR DESCRIPTION
**Pull request by nsulzer**
Attempt to contribute to #2673 to continue the conversation and get some feedback.

Most of it is based on  on #2126, with the addition of forecast based on past averages, and a move to Recahrts. Average spending is based on the average of the relevant month in past years (so the June projection is the average of all June's). If there is no data, it will take the average of all available months.

I am extremely open to feedback. I hope this can help move the discussion along and work towards something good :)

### CODE REVIEW

**Comment by Teprifer** [comment]
> Average spending is based on the average of the relevant month in past years (so the June projection is the average of all June's). If there is no data, it will take the average of all available months.

I'd like to suggest this is limited to the past 1 or 2 Junes, using myself as example, I used Actual before it went free, and imported nYNAB data, I have a number of years of data (back to 2017) and so averaging all those Junes will generate an number not as representative as it could be. I think moving with the last 1 or 2 years is better as personal circumstances change.

Likewise others have imported years of data from nYNAB and YNAB4.

_I lean towards just the last year for my personal circumstances to reflect changes, but see using the last 2 as perfectly valid as well_

**Got an error when trying to add a filter:**
1) Settings:  image

(Same error occurred when using the Average Spending Option)

2) Add category filter per: image

3) Result: 
image

image

_Also recommend checking all the other filter options for the same._

**Comment by mk-french** [comment]
This is great! Thanks for picking it up and contributing! I lost a bit of steam and never really pushed it any further after the PR was closed... The community seemed a little divided on how it should work and with the change in graphing library, I never found the time/motivation.

I did extend it to also use the budget as a source, very similar to how you've done it, see: https://github.com/mk-french/actual/commit/9b2a73c0886632be0d30afe5843131036a37b083 but I hadn't pushed it before the PR was closed.

If I get a moment, I'll try merge it with your changes here. Also I think there may have been a few bugs fixed along the way...

I also echo the issue of including too many years in the past. When testing, my August forecast expense has blown out because of a house purchase from many years back! 😂 
Maybe months with expenses detected as outliers compared to other years could be excluded at a minimum?

There's lot's of work to be done around the filtering, as some of the filtering options don't make sense in some contexts...

**Comment by glowtape** [comment]
I suppose you can't run two filters? One for the current cashflow graph as-is, and another for the forecast based on average spending? Say if I would have had extraordinary expenses that I want in the cashflow graph, but not show in the forecast.

**Comment by nsulzer** [comment]
Hi all, thank you for the testing and suggestions :)

The point of wanting to limit the number of past years considered is a good one. I have (I hope, though I stil struggle to test this) added the option to limit the number of years. That may be a bit overkill, but I think we might as well add features until it becomes clear what sticks.

As to the filtering errors, they should be resolved. However, the filtering options for schedules are limited to account and payee, while date (speaking of filtering options that don't make sense in context...) and amount should be possible as well.

Then to better forecasting, I like both of these ideas.
- > I suppose you can't run two filters? One for the current cashflow graph as-is, and another for the forecast based on average spending? Say if I would have had extraordinary expenses that I want in the cashflow graph, but not show in the forecast. 

  We're already running separate queries, so a spearate filter is definitely possible. It would just require a separate filter selection UI for the forecast.

- > Maybe months with expenses detected as outliers compared to other years could be excluded at a minimum?

  Perhaps instead of removing the outlier month, filtering outlying transaction values using above suggestion might be sufficient. The heuristics around classifying an outlier might be difficult to nail down.

In the trend of throwing features at the wall to see what sticks, I'll definitely have a go at implementing separate filtering for the forecast.  If that doesn't adequately deal with outliers, well... we'll see.

> There's lot's of work to be done around the filtering, as some of the filtering options don't make sense in some contexts...

Definitely, do you have any concrete examples?

**Comment by Teprifer** [comment]
Hi!

I think the average x years isn't working, I have a big lump sum I made in February and so I'd expect the curve to change a lot if I use more years than 1. 
This is may a good way to test too, just have a mega large transaction in a month that is just outside the 1 year point. Then changing between 2 years and 1 should be very obvious.
I think it may be sufficient to have a drop down to set of buttons for 1 year, 2 year, 3 year, and All Time?

_Can see the forecast line looks the same regardless of 1 or 10 years and the big drop in February to affect numbers that more years would have smoothed out_

image
image

Extreme example, all time history, max forecast:

image
image


Regarding filtering, instead of having two sets of filters, maybe instead(if you think it's a good idea, technically doable, etc etc.) might be to simply reset the filters when changing between Average Spending and Schedules (optional toast window to let them know). Then the filter options list itself could be filtered for only what makes sense for each option.

I can kinda see a case for having a different filter list that only applies to the forecasted portion, but even done this way, I think resetting them when changing options is a clean simple way to avoid breakages with incompatible filters and communicate to the user the different availability

In terms of editing the filtering list, Carkom did this in https://github.com/actualbudget/actual/pull/2622 so if you want to(ideally) filter the filter list options based on which forecast is option is used it might help you. :)

**Comment by glowtape** [comment]
See for instance this month November, the expenses are lower than normal and the income is lacking:

image

If I had to guess, since it's beginning of May right now, your code adds six months to the beginning of November, but doesn't round up to the end of the interval, and gets all schedules only until November 5th (it's May 5th posting this).

File: packages/desktop-client/src/components/reports/spreadsheets/cash-flow-spreadsheet.tsx [hunk 1]
```diff
@@ -157,7 +158,13 @@ export function cashFlowByDate(
         })
         .filter({
           $and: [
-            { date: { $transform: '$month', $lte: monthUtils.prevMonth(end) } },
+            {
+              date: {
+                $transform: '$month',
+                $lte: monthUtils.prevMonth(end),
+                $gt: monthUtils.subMonths(end, forecastDivideYears * 12),
```
Comment (mk-french) [inline packages/desktop-client/src/components/reports/spreadsheets/cash-flow-spreadsheet.tsx:165]: I'm not entirely sure how to contribute to your PR (I can't push to your branch in your repo due to permissions) but you need to split these into seperate objects and combine them with an $and operator.
https://github.com/actualbudget/actual/pull/2674#issuecomment-2087988385

Comment (mk-french) [reply packages/desktop-client/src/components/reports/spreadsheets/cash-flow-spreadsheet.tsx:165]: @nsulzer , I have the above fix and also the budget source merged in and ready to push. I've created a PR to your branch.

**Comment by mk-french** [comment]
@nsulzer , would be good to not let this get stale again... Have you got anything further in the pipeline for this? I can take a look at some of the remaining edge cases and then maybe we can get it reviewed for merging?

**Comment by axjasf** [comment]
Could this be reopened? This is an amazing feature.

**Pull request closed without merging**